### PR TITLE
Updated for Session 0.6, made lots of stuff optional

### PR DESCRIPTION
**Pull request by crisward**
Sorry, didn't ready your contribution guidelines until I'd done all this. Hopefully this stuff is of some use.

### CODE REVIEW

File: src/kemal-csrf.cr [hunk 1]
```diff
@@ -10,35 +10,34 @@ require "kemal-session"
 # Without CSRF protection, your app is vulnerable to replay attacks
 # where an attacker can re-submit a form.
 #
-class CSRF < HTTP::Handler
-  HEADER          = "X_CSRF_TOKEN"
-  ALLOWED_METHODS = %w(GET HEAD OPTIONS TRACE)
-  PARAMETER_NAME  = "authenticity_token"
+class CSRF < Kemal::Handler
+  def initialize(@header = "X_CSRF_TOKEN", @allowed_methods = %w(GET HEAD OPTIONS TRACE), @parameter_name = "authenticity_token", @error = "Forbidden")
```
Comment (sdogruyol) [inline src/kemal-csrf.cr:14]: Is there any specific reason to convert these constants to instance variables?

**Comment by crisward** [comment]
I'm building an app which is a drop in replacement for an existing node app. So I needed to set different names for the header and form parameter as the client code is expecting different names. 

It strikes me others may have this need so I made these customisable. As far as I'm aware you can't set constants  within class constructor parameters. Is there a more idiomatic way of achieving this?

File: src/kemal-csrf.cr [hunk 2]
```diff
@@ -10,35 +10,34 @@ require "kemal-session"
...
+                  context.params.body[@parameter_name]
                 else
                   "nothing"
                 end
     current_token = context.session.string("csrf")
 
     if current_token == submitted
       # reset the token so it can't be used again
-      context.session.string("csrf", SecureRandom.hex(16))
+      # context.session.string("csrf", SecureRandom.hex(16))
```
Comment (sdogruyol) [inline src/kemal-csrf.cr:36]: Why is this commented out?

**Comment by sdogruyol** [comment]
@crisward aha! Now i got why you made those ivars. Makes sense 👍

File: src/kemal-csrf.cr (hunk 2 above)
Comment (crisward) [reply src/kemal-csrf.cr:36]: https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet

This article suggests on csrf per session. It also helps with the back button and using csrf tokens with client side javascript.


> In general, developers need only generate this token once for the current session. After initial generation of this token, the value is stored in the session and is utilized for each subsequent request until the session expires. When a request is issued by the end-user, the server-side component must verify the existence and validity of the token in the request as compared to the token found in the session. If the token was not found within the request or the value provided does not match the value within the session, then the request should be aborted, token should be reset and the event logged as a potential CSRF attack in progress.

Comment (crisward) [reply src/kemal-csrf.cr:36]: It actually suggests resetting the token if it is **not** valid. So that if could have an else which resets the token.

**Comment by crisward** [comment]
Thanks!

**Pull request merged**